PR: https://github.com/odoo/odoo/pull/73174

From: 4cd70f164ea1400925dc8237e92411b304c4f990
From: Sébastien Theys
Date: 2021-08-12 15:22:59

Structural Changes: 2
Total Changes: 46

[IMP] mail, hr, hr_holidays, *: batch mail partner format

* = test_discuss_full

Overall less queries.

Changes:

- Batch `mail_partner_format`.
- Batch `user.employee_id` compute.
- Remove query for current leave state, use fields (and prefetch) instead.
- Move current leave state in partner format instead of channel info.

Part of task-2622462

closes odoo/odoo#73174

Related: odoo/enterprise#20186
Signed-off-by: Sébastien Theys (seb) <seb@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays/models/__init__.py
+++ b/addons/hr_holidays/models/__init__.py
@@ -7,7 +7,6 @@ from . import hr_department
 from . import hr_leave
 from . import hr_leave_allocation
 from . import hr_leave_type
-from . import mail_channel
 from . import mail_message_subtype
 from . import res_partner
 from . import res_users

--- a/addons/hr_holidays/models/mail_channel.py
+++ b/None
@@ -1,32 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import fields, models
-
-
-class Channel(models.Model):
-    _inherit = 'mail.channel'
-
-    def _get_channel_partner_info(self, all_partners, direct_partners):
-        partner_infos = super(Channel, self)._get_channel_partner_info(all_partners, direct_partners)
-        # only search for leave out_of_office_date_end if im_status is on leave
-        partners_on_leave = [partner_id for partner_id in direct_partners.ids if 'leave' in partner_infos[partner_id]['im_status']]
-        if partners_on_leave:
-            now = fields.Datetime.now()
-            self.env.cr.execute('''SELECT res_users.partner_id as partner_id, hr_leave.date_to as date_to
-                                FROM res_users
-                                JOIN hr_leave ON hr_leave.user_id = res_users.id
-                                AND hr_leave.state not in ('cancel', 'refuse')
-                                AND res_users.active = 't'
-                                AND hr_leave.date_from <= %s
-                                AND hr_leave.date_to >= %s
-                                AND res_users.partner_id in %s''', (now, now, tuple(partners_on_leave)))
-            out_of_office_infos = dict(((res['partner_id'], res) for res in self.env.cr.dictfetchall()))
-            for partner_id, out_of_office_info in out_of_office_infos.items():
-                partner_infos[partner_id]['out_of_office_date_end'] = out_of_office_info['date_to']
-
-        # fill empty values for the consistency of the result
-        for partner_info in partner_infos.values():
-            partner_info.setdefault('out_of_office_date_end', False)
-
-        return partner_infos

--- a/addons/hr_holidays/models/res_partner.py
+++ b/addons/hr_holidays/models/res_partner.py
@@ -2,6 +2,7 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, models
+from odoo.tools.misc import DEFAULT_SERVER_DATE_FORMAT
 
 
 class ResPartner(models.Model):
@@ -22,3 +23,15 @@ class ResPartner(models.Model):
     @api.model
     def _get_on_leave_ids(self):
         return self.env['res.users']._get_on_leave_ids(partner=True)
+
+    def mail_partner_format(self):
+        """Override to add the current leave status."""
+        partners_format = super().mail_partner_format()
+        for partner in self:
+            # in the rare case of multi-user partner, return the earliest possible return date
+            dates = sorted(partner.mapped('user_ids.leave_date_to'))
+            date = dates[0] if dates and all(dates) else False
+            partners_format.get(partner).update({
+                'out_of_office_date_end': date.strftime(DEFAULT_SERVER_DATE_FORMAT) if date else False,
+            })
+        return partners_format
